R/position-sina.r

```diff
@@ -0,0 +1,221 @@
+    trans_x[[j]] <- data.frame( "x" = x[keep],
+                                "y" = data$y[keep],
+                                "idx" = idx[keep])
+
+    if (sum(keep) < 2)
+      next
+
+    #per bin sample count
+    neighbours[[j]] <- table(findInterval(trans_x[[j]]$y, bins))
+
+    #find the densiest neighbourhood in the current group and compare it
+    #with the global max
+    tmp_max_neighbours <- max(neighbours[[j]])
+
+    if (tmp_max_neighbours > max_neighbours)
+      max_neighbours <- tmp_max_neighbours
+
+    if (method == "density") {
+
+      #per bin sample density
+      densities[[j]] <- stats::density(trans_x[[j]]$y, adjust = adjust)
+
+      #find the highest density value
+      tmp_max_density <- max(densities[[j]]$y)
+
+      if (tmp_max_density > max_density)
+        max_density <- tmp_max_density
+
+    }
+  }
+
+  # -------------------------------------------------------------------------- #
+
+  for (j in levels(groups)) {
+
+    if (nrow(trans_x[[j]]) < 2 ) next
+
+    #confine the samples in a (-0.5, 0.5) area around the class center
+    if (method == "density") {
+      if (max(densities[[j]]$y) > 0.48)
+        global_scaling_factor <- 0.48 / max(densities[[j]]$y)
+      else
+        global_scaling_factor <- 1
+
+    } else {
+      #if the space required to spread the samples in a neighbourhood exceeds
+      #1, create  compress the points
+      if (max(neighbours[[j]]) > 1 / adjust) {
+        global_scaling_factor <- (1 / adjust) / max(neighbours[[j]])
+      } else
+        global_scaling_factor <- 1
+    }
+
+    #scale all neighbourhoods based on their density relative to the
+    #densiest neighbourhood
+    if (scale == TRUE)
+      group_scaling_factor <- max(neighbours[[j]]) / max_neighbours
+    else
+      group_scaling_factor <- 1
+
+    for (i in names(neighbours[[j]])) {
+
+      #examine neighbourhoods with more than 'neighbour_limit' samples
+      if (neighbours[[j]][i] > neighbour_limit){
+        cur_bin <- bins[ as.integer(i) : (as.integer(i) + 1)]
+
+        #find samples in the current bin and translate their X coord.
+        points <- findInterval(trans_x[[j]]$y, cur_bin) == 1
+
+        #compute the border margin for the current bin
+        if (method == "density")
+          xmax <- mean(densities[[j]]$y[findInterval(densities[[j]]$x,
+                                                     cur_bin) == 1])
+        else
+          xmax <- adjust * neighbours[[j]][i] / 2
+
+        #assign the samples uniformely within the specified range
+        x_translation <- stats::runif(neighbours[[j]][i], - xmax, xmax )
+
+        #scale and store new x coordinates
+        trans_x[[j]]$x[points] <- trans_x[[j]]$x[points] +
+          (x_translation * global_scaling_factor * group_scaling_factor)
+      }
+    }
+  }
+
+  #collapse list to data frame
+  trans_x <- do.call(rbind, trans_x)
+  #return only the transposed x values in input order
+  trans_x$x[order(trans_x$idx)]
```

Comment:
For the `stat$compute_panel` method, you would return the `trans_x` dataframe with the `x` and `y` values. I would delete `idx` because it would not be needed any more. However, find a way to include the `group` in there.

Maybe this would do it

```R
"group" = data$group[keep],
```